java/ql/lib/semmle/code/java/security/SuspiciousRegexpRangeQuery.qll

```diff
@@ -0,0 +1,302 @@
+/**
+ * Classes and predicates for working with suspicious character ranges.
+ */
+
+// We don't need the NFA utils, just the regexp tree.
+// but the below is a nice shared library that exposes the API we need.
+import performance.ReDoSUtil
+
+/**
+ * Gets a rank for `range` that is unique for ranges in the same file.
+ * Prioritizes ranges that match more characters.
+ */
+int rankRange(RegExpCharacterRange range) {
+  range =
+    rank[result](RegExpCharacterRange r, Location l, int low, int high |
+      r.getLocation() = l and
+      isRange(r, low, high)
+    |
+      r order by (high - low) desc, l.getStartLine(), l.getStartColumn()
+    )
+}
+
+/** Holds if `range` spans from the unicode code points `low` to `high` (both inclusive). */
+predicate isRange(RegExpCharacterRange range, int low, int high) {
+  exists(string lowc, string highc |
+    range.isRange(lowc, highc) and
+    low.toUnicode() = lowc and
+    high.toUnicode() = highc
+  )
+}
+
+/** Holds if `char` is an alpha-numeric character. */
+predicate isAlphanumeric(string char) {
+  // written like this to avoid having a bindingset for the predicate
+  char = [[48 .. 57], [65 .. 90], [97 .. 122]].toUnicode() // 0-9, A-Z, a-z
+}
+
+/**
+ * Holds if the given ranges are from the same character class
+ * and there exists at least one character matched by both ranges.
+ */
+predicate overlap(RegExpCharacterRange a, RegExpCharacterRange b) {
+  exists(RegExpCharacterClass clz |
+    a = clz.getAChild() and
+    b = clz.getAChild()
+  |
+    // b contains the lower end of a
+    exists(int alow, int ahigh, int blow, int bhigh |
+      isRange(a, alow, ahigh) and
+      isRange(b, blow, bhigh) and
+      blow <= alow and
+      bhigh >= ahigh
+    )
+    or
+    // b contains the upper end of a
+    exists(int blow, int bhigh, int alow, int ahigh |
+      isRange(a, alow, ahigh) and
+      isRange(b, blow, bhigh) and
+      blow <= ahigh and
+      bhigh >= ahigh
+    )
+  )
+  or
+  // symmetric overlap
+  overlap(b, a)
```

Comment:
These 3 cases can be simplified to a single case: `alow <= bhigh and blow <= ahigh`. To realise this consider the two cases for non-overlap: `alow > bhigh or blow > ahigh` - the simplified condition is merely the negation of that.
(also there's a mistake in line 52, but that doesn't matter when all this is simplified.
```suggestion
    exists(int alow, int ahigh, int blow, int bhigh |
      isRange(a, alow, ahigh) and
      isRange(b, blow, bhigh) and
      alow <= bhigh and
      blow <= ahigh
    )
  )
```

Reply:
Also, should we have `a != b`?

Reply:
> These 3 cases can be simplified to a single case: `alow <= bhigh and blow <= ahigh`. To realise this consider the two cases for non-overlap: `alow > bhigh or blow > ahigh` - the simplified condition is merely the negation of that.

I had to visualize it using my left/right hands as the a/b ranges, but now I get it 👍  
That makes it way simpler, thanks.